### PR TITLE
BCI-3668 Optimise HeadTracker's memory usage

### PR DESCRIPTION
**Pull request by dhaidashenko**
[Ticket.](https://smartcontract-it.atlassian.net/browse/BCI-3668)

Previously HeadTracker performed deep copy of the canonical chain every time new block was added to avoid race condition on .Parent field, when old blocks were removed. This causes large number of allocations and as result increases CPU usage due to GC.

PR replaces deep copy with atomic pointer, which results in slower iteration through the chain, but significantly reduces number of allocations. 
Benchmark results:
Iterate through whole chain:
```
benchstat lock_free_earliest_head_in_chain atomic_earliest_head_in_chain
goos: darwin
goarch: arm64
pkg: github.com/smartcontractkit/chainlink/v2/core/chains/evm/headtracker
                       │ lock_free_earliest_head_in_chain │   atomic_earliest_head_in_chain    │
                       │              sec/op              │   sec/op     vs base               │
EarliestHeadInChain-10                        15.21m ± 6%   15.99m ± 4%  +5.12% (p=0.005 n=10)

                       │ lock_free_earliest_head_in_chain │ atomic_earliest_head_in_chain  │
                       │               B/op               │    B/op     vs base            │
EarliestHeadInChain-10                         72.00 ± 0%   72.00 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal

                       │ lock_free_earliest_head_in_chain │ atomic_earliest_head_in_chain  │
                       │            allocs/op             │ allocs/op   vs base            │
EarliestHeadInChain-10                         3.000 ± 0%   3.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```

Add new head to the chain:
```
 benchstat lockfee_backfill atomic_backfill
goos: darwin
goarch: arm64
pkg: github.com/smartcontractkit/chainlink/v2/core/chains/evm/headtracker
                           │ lockfee_backfill │           atomic_backfill           │
                           │      sec/op      │   sec/op     vs base                │
Heads_SimulatedBackfill-10    4694883.0n ± 1%   801.4n ± 1%  -99.98% (p=0.000 n=10)

                           │ lockfee_backfill │          atomic_backfill           │
                           │       B/op       │    B/op     vs base                │
Heads_SimulatedBackfill-10    13295246.5 ± 0%   713.0 ± 0%  -99.99% (p=0.000 n=10)

                           │ lockfee_backfill │          atomic_backfill           │
                           │    allocs/op     │ allocs/op   vs base                │
Heads_SimulatedBackfill-10     32057.000 ± 0%   6.000 ± 0%  -99.98% (p=0.000 n=10)
```

### CODE REVIEW

File: core/chains/evm/types/models.go [hunk 1]
```diff
@@ -90,10 +94,49 @@ func (h *Head) BlockDifficulty() *big.Int {
...
+	var earliestInChain *Head
+	h.forEach(func(h *Head) bool {
+		earliestInChain = h
+		return true
+	})
+	return earliestInChain
+}
+
+// forEach - go through the chain linked list. Stops when do returns false or h is nil
+func (h *Head) forEach(do func(h *Head) bool) {
```
Comment (jmank88) [inline core/chains/evm/types/models.go]: We could use the new for-range `iter.Seq` API here. Your pattern is nearly identical:
```suggestion
// forEach - go through the chain linked list. Stops when do returns false or h is nil
func (h *Head) all() iter.Seq[*Head] {
	return func(yield func(*Head) bool) {
		...
		if !yield(cur) {
			return
		}
```
Then the callers can do regular for loops:
```diff
- h.forEach(func(h *Head) bool {
+ for _ := range h.all() {
```

File: core/chains/evm/types/models.go [hunk 2]
```diff
@@ -90,10 +94,49 @@ func (h *Head) BlockDifficulty() *big.Int {
...
+		cur = cur.Parent.Load()
+
+		// if there is a circular reference eventually `fast` will catch up with `cur`.
+		// As .Parent can be modified by another goroutine, it's possible that
+		// cycle will be created after `fast` reached end of the list. To prevent infinite loop, we use duration based validation
+		if (cur != nil && cur == fast) || time.Since(start) > 10*time.Second {
+			// we should log the issue instead of panicking to avoid killing whole node.
+			// we can switch to panic, after full transition to loops.
+			lggr, _ := logger.New()
+			logger.Sugared(lggr).Criticalw("Detected cycle in chain structure. Aborting iteration.",
```
Comment (jmank88) [inline core/chains/evm/types/models.go]: We cannot construct new loggers arbitrarily. You will either have to inject it, or return errors to be logged from a higher level.

File: core/chains/evm/types/models.go (hunk 1 above)
Comment (jmank88) [reply core/chains/evm/types/models.go]: https://github.com/smartcontractkit/chainlink/pull/14353

File: core/chains/evm/types/models.go (hunk 2 above)
Comment (dhaidashenko) [reply core/chains/evm/types/models.go]: Hmm, we do cycle detection when actually setting `Parent`, so these checks are not required.
Returning an error will break many method signatures, and injecting the logger does not seem like a clean solution either.
Dropping the check seems like the best option.

File: core/chains/evm/types/models.go (hunk 1 above)
Comment (dhaidashenko) [reply core/chains/evm/types/models.go]: The iterator is no longer needed, as I've dropped the cycle detection (the reason is in the comment below).

**Review by jmank88 — DISMISSED** [review]

File: core/chains/evm/headtracker/heads.go [hunk 3]
```diff
@@ -26,114 +27,158 @@ type Heads interface {
...
+	}
+}
+
+func (h *heads) ensureNoCycles(newHead *evmtypes.Head) error {
+	if newHead.ParentHash == newHead.Hash {
+		return fmt.Errorf("cycle detected: newHeads reference itself newHead(%s)", newHead.String())
+	}
+	if parent, ok := h.headsByHash[newHead.ParentHash]; ok {
+		if parent.Number >= newHead.Number {
+			return fmt.Errorf("potential cycle detected while adding newHead as child: %w", newPotentialCycleError(parent, newHead))
```
Comment (dimriou) [inline core/chains/evm/headtracker/heads.go:120]: I understand this is a cycle problem in the context of Head Tracker, but this is more of a general problem as the parent's block number is higher than the child that references it. I would consider this as invalid head/data from the chain.

File: core/chains/evm/headtracker/heads.go [hunk 4]
```diff
@@ -26,114 +27,158 @@ type Heads interface {
...
-	for finalizedHead != nil {
-		finalizedHead.IsFinalized = true
-		finalizedHead = finalizedHead.Parent
+
+	markFinalized(finalizedHead)
+
+	// remove all blocks that are older than minBlockToKeep
+	for h.headsAsc.Len() > 0 && h.headsAsc.Peek().Number < minBlockToKeep {
+		oldBlock := heap.Pop(h.headsAsc).(*evmtypes.Head)
+		delete(h.headsByHash, oldBlock.Hash)
```
Comment (dimriou) [inline core/chains/evm/headtracker/heads.go:90]: Would it be possible to delete and return a copy within `heap.Pop` instead?

File: core/chains/evm/headtracker/heap.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	h.values[i], h.values[j] = h.values[j], h.values[i]
+}
+
+func (h *headsHeap) Less(i, j int) bool {
+	return h.values[i].Number < h.values[j].Number
+}
+
+func (h *headsHeap) Pop() any {
+	old := h.values[len(h.values)-1]
+	h.values = h.values[:len(h.values)-1]
```
Comment (dimriou) [inline core/chains/evm/headtracker/heap.go]: Given the heap is structured in ascending order, shouldn't `old` pop the smallest/first item of the array?

Comment (dhaidashenko) [reply core/chains/evm/headtracker/heap.go]: `headsHeap` implements `heap.Interface`, which requires Pop to remove the last element from the slice.

File: core/chains/evm/headtracker/heads.go (hunk 4 above)
Comment (dhaidashenko) [reply core/chains/evm/headtracker/heads.go:90]: Not sure that I understand. Could you elaborate?

File: core/chains/evm/headtracker/heads.go (hunk 3 above)
Comment (dhaidashenko) [reply core/chains/evm/headtracker/heads.go:120]: `newPotentialCycleError` already signals that we expect block number `to strictly decrease in 'child -> parent' relation`.  Do you suggest changing this msg?

File: core/chains/evm/headtracker/heads.go (hunk 4 above)
Comment (dimriou) [reply core/chains/evm/headtracker/heads.go:90]: I thought you were dereferencing the popped item from the Heap here, which leads me to the actual question I had: should we do this `h.values[len(h.values)-1] = nil` inside Pop to avoid memory leaks?

File: core/chains/evm/headtracker/heads.go (hunk 3 above)
Comment (dimriou) [reply core/chains/evm/headtracker/heads.go:120]: `newPotentialCycleError` error message is accurate. Perhaps I would rephrase the `potential cycle...` part because if the hashes are right, from a user perspective the issue would be better categorized as bad data on the block number rather than a cycle problem, but it's just me nitpicking. You can ignore this.

File: core/chains/evm/types/models.go [hunk 6]
```diff
@@ -142,47 +139,29 @@ func (h *Head) ChainLength() uint32 {
 	if h == nil {
 		return 0
 	}
```
Comment (dimriou) [inline core/chains/evm/types/models.go]: Should we remove this since you already check for nil pointer below?

**Review by dimriou — DISMISSED** [review]

**Review by DylanTinianov — DISMISSED** [review]

File: core/chains/evm/headtracker/heap.go
Comment (jmank88) [inline core/chains/evm/headtracker/heap.go:1]: This is fine, but I wonder if we have a generic heap type anywhere already.

**Review by jmank88 — DISMISSED** [review]

**Review by jmank88 — APPROVED** [review]

**Review by dimriou — APPROVED** [review]
